invenio_app_ils/records_relations/retriever.py

```diff
@@ -0,0 +1,239 @@
+        r.update(relevant_fields)
+        return r
+
+    def get(self):
+        """Return only relations that this record has with parents.
+
+        Given that in ILS system parent records (e.g. journal, periodicals)
+        can have a very high number of children but only a few parents,
+        only relations with the parents are returned.
+        """
+        relations = {}
+        for relation_type in current_app.config["PARENT_CHILD_RELATION_TYPES"]:
+            pcr = ParentChildRelation(relation_type)
+            name = relation_type.name
+
+            for parent_pid in pcr.get_parents_of(self.child_record.pid):
+                r = self._build_relation_obj(
+                    self.child_record, parent_pid, name
+                )
+                relations.setdefault(name, [])
+                relations[name].append(r)
+        return relations
+
+
+class SiblingsRetriever(RelationObjectBuilderMixin):
+    """Retrieve relations of type Siblings."""
+
+    def __init__(self, record):
+        """Constructor."""
+        self.record = record
+
+    @staticmethod
+    def _get_extra_metadata(record, sibling, relation_type):
+        """Retrieve any extra metadata in the current record or sibling."""
+        record_pid = record.pid
+        sibling_pid = sibling.pid
+        # we don't know it extra metadata are stored in this record or in the
+        # sibling. Need to check both.
+        metadata = {}
+
+        # get any extra metadata stored in this record for the relation with
+        # the sibling
+        rec_metadata = RelationsExtraMetadata.get_extra_metadata_from(
+            record, relation_type, sibling_pid.pid_value, sibling_pid.pid_type
+        )
+        metadata.update(rec_metadata)
+
+        # get any extra metadata stored in the sibling for the relation with
+        # this record
+        sibling_metadata = RelationsExtraMetadata.get_extra_metadata_from(
+            sibling, relation_type, record_pid.pid_value, record_pid.pid_type
+        )
+        if (
+            sibling_metadata
+            and "pid_value" in sibling_metadata
+            and "pid_type" in sibling_metadata
+        ):
+            # the extra metadata in the sibling has pid_value and pid_type of
+            # this record. Replace it with the sibling ones.
+            sibling_metadata["pid_value"] = sibling_pid.pid_value
+            sibling_metadata["pid_type"] = sibling_pid.pid_type
+
+        metadata.update(sibling_metadata)
+        return metadata
+
+    def _build_relation_obj(self, sibling_pid, relation_type):
+        """Return the relation object with metadata."""
+        pid_value = sibling_pid.pid_value
+        pid_type = sibling_pid.pid_type
+
+        r = self.build_relations_object(pid_value, pid_type, relation_type)
+
+        # fetch the sibling to retrieve useful metadata (title, ...) and also
+        # any optional extra metadata for this relation
+        from invenio_app_ils.records.api import IlsRecord
```

Comment:
Can we move this on top?

Reply:
As Karolina said, circular dep otherwise